dap-estgi-server/src/Main.hs

```diff
@@ -486,8 +565,23 @@ talk CommandScopes = do
 ----------------------------------------------------------------------------
 talk CommandVariables = do
   VariablesArguments {..} <- getArguments
-  variables <- getVariables variablesArgumentsVariablesReference
-  sendVariablesResponse (VariablesResponse variables)
+  getsApp (Bimap.lookupR variablesArgumentsVariablesReference . dapVariablesRefMap) >>= \case
+    Just VariablesRef_StackFrameVariables{} -> do
+      variables <- getVariables variablesArgumentsVariablesReference
+      sendVariablesResponse (VariablesResponse variables)
+    Just (VariablesRef_HeapObject addr) -> do
+      stgState <- getStgState
+      let Just ho = IntMap.lookup addr $ ssHeap stgState
+      variables <- getVariablesForHeapObject stgState ho
+      -- detect and annotate loops
+      let markLoop v
+            | variableVariablesReference v == 0
+            = v
+            | variableVariablesReference v > variablesArgumentsVariablesReference
+            = v
+            | otherwise
+            = v {variableName = variableName v <> " <loop>"}
```

Comment:
Consider different naming for `<loop>`, reminds me of an infinite loop detection, could potentially cause confusion for others.